ferry/deploy.py

```diff
@@ -185,61 +220,100 @@ def _match(name):
         revert = conn.begin()
         conn.execute("SET CONSTRAINTS ALL DEFERRED;")
 
-        for table in meta.sorted_tables[::-1]:
-            if not table.name.endswith("_staged"):
-                print(f"Reverting table {table.name}")
-                conn.execute(
-                    f'ALTER TABLE "{table.name}" RENAME TO {table.name}_staged;'
-                )
-                conn.execute(f'ALTER TABLE "{table.name}_old" RENAME TO {table.name};')
+        for table in alchemy_tables:
+            print(f"Reverting table {table.name}")
+            conn.execute(
+                f'ALTER TABLE IF EXISTS "{table.name}" RENAME TO {table.name}_staged;'
+            )
+            conn.execute(
+                f'ALTER TABLE IF EXISTS "{table.name}_old" RENAME TO {table.name};'
+            )
 
         revert.commit()
         raise
 
+    # -----------------
+    # Remove old tables
+    # -----------------
+
     print("\n[Deleting temporary old tables]")
-    meta.reflect()  # update the models to reflect postgres
     delete = conn.begin()
     conn.execute("SET CONSTRAINTS ALL DEFERRED;")
-    for table in meta.sorted_tables[::-1]:
-        if table.name.endswith("_old"):
-            print(f"Dropping table {table.name}")
-            conn.execute(f"DROP TABLE IF EXISTS {table.name} CASCADE;")
+    for table in alchemy_tables:
+        print(f"Dropping table {table.name}_old")
+        conn.execute(f"DROP TABLE IF EXISTS {table.name}_old CASCADE;")
 
     delete.commit()
 
+    # ---------------
+    # Add constraints
+    # ---------------
+
     # add back the foreign key constraints
-    print("\n[Regenerating constraints]")
+    print(f"\n[Regenerating constraints ({len(constraints)})]")
     regen_constraints = conn.begin()
     for constraint in constraints:
         if isinstance(constraint, ForeignKeyConstraint):
             conn.execute(schema.AddConstraint(constraint))
     regen_constraints.commit()
 
+    # --------------
+    # Create indexes
+    # --------------
+
     def index_exists(name):
         result = conn.execute(
             f"SELECT exists(SELECT 1 from pg_indexes where indexname = '{name}') as ix_exists;"
         ).first()
         return result.ix_exists
 
-    print("\n[Regenerating indexes]")
+    print(f"\n[Regenerating indexes ({len(indexes)})]")
     regen_indexes = conn.begin()
     for index in indexes:
         if index_exists(index.name):
             conn.execute(schema.DropIndex(index))
         conn.execute(schema.CreateIndex(index))
     regen_indexes.commit()
```

Comment:
hmm can we not generate the constraints and indexes on the _staged tables?

Reply:
That was my initial plan in the importer, but SQLAlchemy tries to generate them connecting back to the other main tables (particularly for foreign key constraints), which raises an error.

Reply:
Updated so that all constraints excepting foreign keys are generated on staged tables, and only foreign key constraints are added after deployment. Indexes are tricky because SQLAlchemy tries to generate them for the staged tables automatically, but names then with _staged.